### PR TITLE
Introduce TrajectoryBuilderInterface.

### PR DESCRIPTION
**Pull request by cschuet**
[RFC=0004](https://github.com/googlecartographer/rfcs/blob/master/text/0004-trajectory-builder-interface.md)

### CODE REVIEW

File: cartographer/mapping/map_builder_interface.h [hunk 1]
```diff
@@ -61,8 +61,8 @@ class MapBuilderInterface {
   // Returns the TrajectoryBuilder corresponding to the specified
   // 'trajectory_id' or 'nullptr' if the trajectory has no corresponding
   // builder.
-  virtual mapping::TrajectoryBuilder* GetTrajectoryBuilder(
-      int trajectory_id) const = 0;
+  virtual mapping::GlobalTrajectoryBuilderInterface *
+  GetTrajectoryBuilder(int trajectory_id) const = 0;
```
Comment (gaschler) [inline cartographer/mapping/map_builder_interface.h]: The method name can stay the same if we are planning to drop "Global" from GlobalTrajectoryBuilderInterface. Otherwise, rename.

File: cartographer/mapping/map_builder_interface.h [hunk 2]
```diff
@@ -61,8 +61,8 @@ class MapBuilderInterface {
   // Returns the TrajectoryBuilder corresponding to the specified
```
Comment (gaschler) [inline cartographer/mapping/map_builder_interface.h]: update comment

File: cartographer/mapping/collated_trajectory_builder.h [hunk 3]
```diff
@@ -49,33 +49,29 @@ class CollatedTrajectoryBuilder : public TrajectoryBuilder {
   CollatedTrajectoryBuilder& operator=(const CollatedTrajectoryBuilder&) =
       delete;
 
-  void AddRangefinderData(const std::string& sensor_id, common::Time time,
-                          const Eigen::Vector3f& origin,
-                          const sensor::TimedPointCloud& ranges) override {
+  void AddSensorData(
+      const std::string &sensor_id,
+      const sensor::TimedPointCloudData &timed_point_cloud_data) override {
     AddSensorData(sensor_id,
```
Comment (gaschler) [inline cartographer/mapping/collated_trajectory_builder.h]: drop this argument "sensor_id"

File: cartographer/internal/mapping/global_trajectory_builder.h [hunk 4]
```diff
@@ -42,12 +42,15 @@ class GlobalTrajectoryBuilder
   GlobalTrajectoryBuilder(const GlobalTrajectoryBuilder&) = delete;
   GlobalTrajectoryBuilder& operator=(const GlobalTrajectoryBuilder&) = delete;
 
-  void AddRangefinderData(const common::Time time,
-                          const Eigen::Vector3f& origin,
-                          const sensor::TimedPointCloud& ranges) override {
+  void AddSensorData(
+      const std::string &sensor_id,
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/internal/mapping/global_trajectory_builder.h]: Done.

File: cartographer/mapping/collated_trajectory_builder.h (hunk 3 above)
Comment (cschuet) [reply cartographer/mapping/collated_trajectory_builder.h]: Done.

File: cartographer/mapping/map_builder_interface.h (hunk 2 above)
Comment (cschuet) [reply cartographer/mapping/map_builder_interface.h]: Done.

File: cartographer/mapping/map_builder_interface.h (hunk 1 above)
Comment (cschuet) [reply cartographer/mapping/map_builder_interface.h]: Sounds good.

File: cartographer/mapping/trajectory_builder.h [hunk 5]
```diff
@@ -37,32 +37,6 @@ namespace mapping {
 proto::TrajectoryBuilderOptions CreateTrajectoryBuilderOptions(
     common::LuaParameterDictionary* const parameter_dictionary);
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/mapping/trajectory_builder.h]: Obsolete.

File: cartographer/mapping/trajectory_builder.h [hunk 6]
```diff
@@ -37,32 +37,6 @@ namespace mapping {
...
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/mapping/trajectory_builder.h]: It does. PR pending.

File: cartographer/sensor/collator.h [hunk 7]
```diff
@@ -51,8 +51,7 @@ class Collator {
   // Adds 'data' for 'trajectory_id' to be collated. 'data' must contain valid
   // sensor data. Sensor packets with matching 'sensor_id' must be added in time
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/sensor/collator.h]: Done.

File: cartographer/sensor/data.h [hunk 8]
```diff
@@ -31,50 +31,38 @@ namespace sensor {
 
 class Data {
  public:
-  virtual ~Data() {}
+   Data(const std::string &sensor_id) : sensor_id_(sensor_id) {}
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/sensor/data.h]: Done.

File: cartographer/sensor/data.h [hunk 9]
```diff
@@ -31,50 +31,38 @@ namespace sensor {
...
-  const TimedPointCloud ranges_;
+ protected:
+   const std::string sensor_id_;
 };
 
 template <typename DataType>
 class Dispatchable : public Data {
  public:
-  Dispatchable(const DataType& data) : data_(data) {}
+   Dispatchable(const std::string &sensor_id, const DataType &data)
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/sensor/data.h]: Ack.

File: cartographer/sensor/timed_point_cloud_data.h [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+#define CARTOGRAPHER_SENSOR_TIMED_POINT_CLOUD_DATA_H_
+
+#include "Eigen/Core"
+#include "cartographer/common/time.h"
+#include "cartographer/sensor/point_cloud.h"
+
+namespace cartographer {
+namespace sensor {
+
+struct TimedPointCloudData {
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/sensor/timed_point_cloud_data.h]: As discussed we leave this as is.

File: cartographer/mapping/collated_trajectory_builder.h [hunk 11]
```diff
@@ -36,13 +35,12 @@ namespace mapping {
 
 // Handles collating sensor data using a sensor::Collator, then passing it on to
 // a mapping::GlobalTrajectoryBuilderInterface which is common for 2D and 3D.
```
Comment (gaschler) [inline cartographer/mapping/collated_trajectory_builder.h]: update comment

File: cartographer/sensor/data.h [hunk 12]
```diff
@@ -46,13 +44,13 @@ class Data {
...
+  Dispatchable(const std::string &sensor_id, const DataType &data)
+      : Data(sensor_id), data_(data) {}
 
-   common::Time GetTime() const override { return data_.time; }
-   void AddToTrajectoryBuilder(mapping::GlobalTrajectoryBuilderInterface
-                                   *const trajectory_builder) override {
-     trajectory_builder->AddSensorData(sensor_id_, data_);
+  common::Time GetTime() const override { return data_.time; }
+  void AddToTrajectoryBuilder(
+      mapping::TrajectoryBuilderInterface *const trajectory_builder) override {
```
Comment (gaschler) [inline cartographer/sensor/data.h]: style: mapping::TrajectoryBuilderInterface* const
also above

File: cartographer/mapping/collated_trajectory_builder.h (hunk 11 above)
Comment (cschuet) [reply cartographer/mapping/collated_trajectory_builder.h]: Done.

File: cartographer/mapping/trajectory_builder_interface.h [hunk 13]
```diff
@@ -69,4 +75,4 @@ class GlobalTrajectoryBuilderInterface {
 }  // namespace mapping
```
Comment (gaschler) [inline cartographer/mapping/trajectory_builder_interface.h]: Drive-by: style: type declaration (the callback) comes before all other declarations.

File: cartographer/mapping/collated_trajectory_builder.cc [hunk 14]
```diff
@@ -31,8 +31,7 @@ constexpr double kSensorDataRatesLoggingPeriodSeconds = 15.;
 CollatedTrajectoryBuilder::CollatedTrajectoryBuilder(
     sensor::Collator* const sensor_collator, const int trajectory_id,
     const std::unordered_set<std::string>& expected_sensor_ids,
-    std::unique_ptr<GlobalTrajectoryBuilderInterface>
-        wrapped_trajectory_builder)
+    std::unique_ptr<TrajectoryBuilderInterface> wrapped_trajectory_builder)
```
Comment (gaschler) [inline cartographer/mapping/collated_trajectory_builder.cc]: consistency: usually we don't name unique_ptrs "wrapped_"
rename to: global_trajectory_builder, also the member

Comment (cschuet) [reply cartographer/mapping/collated_trajectory_builder.cc]: So "wrapped" does not refer to the fact that it is wrapped in a unique_ptr but rather that it is the `CollatedTrajectoryBuilder` that wraps it. I wouldn't call it `global_trajectory_builder` because in principle a `CollatedTrajectoryBuilder` can wrap any implementation of `TrajectoryBuilderInterface`.

File: cartographer/mapping/trajectory_builder_interface.h (hunk 13 above)
Comment (cschuet) [reply cartographer/mapping/trajectory_builder_interface.h]: Done.

File: cartographer/sensor/collator.h [hunk 15]
```diff
@@ -49,10 +49,9 @@ class Collator {
   void FinishTrajectory(int trajectory_id);
 
   // Adds 'data' for 'trajectory_id' to be collated. 'data' must contain valid
-  // sensor data. Sensor packets with matching 'sensor_id' must be added in time
+  // sensor data. Sensor packets with matching sensor ID must be added in time
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/sensor/collator.h]: Done.

File: cartographer/sensor/data.h [hunk 16]
```diff
@@ -19,11 +19,7 @@
 
 #include "cartographer/common/make_unique.h"
 #include "cartographer/common/time.h"
-#include "cartographer/mapping/global_trajectory_builder_interface.h"
-#include "cartographer/mapping/trajectory_builder.h"
-#include "cartographer/sensor/fixed_frame_pose_data.h"
-#include "cartographer/sensor/imu_data.h"
-#include "cartographer/sensor/odometry_data.h"
+#include "cartographer/mapping/trajectory_builder_interface.h"
 #include "cartographer/sensor/point_cloud.h"
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/sensor/data.h]: Done.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.
